### PR TITLE
SNOW-3511808: Escape single quotes in subfieldExpression and singleQuote

### PR DESCRIPTION
**Pull request by sfc-gh-helmeleegy**
1. What Jira ticket or GitHub issue is this PR addressing? Make sure that there is a ticket or issue accompanying your PR.

   Fixes SNOW-3511808

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   *Escape single quotes in subfieldExpression and singleQuote.*

### CODE REVIEW

**Comment by sfc-gh-yuwang** [comment]
1. collateExpression call site is unprotected — same injection class
                                                                                                                                                                                                                                                                                                                                                                                                                                           
  collateExpression in SqlGenerator.scala:258 calls singleQuote(collationSpec) where collationSpec comes from the user-supplied Column.collate(String) API. After this PR, singleQuote now properly escapes on the wrapping branch — so this is actually fixed as a side effect. But it's worth explicitly noting this in the PR description, and adding a test:                                                                           
                                                                                                                                                                                                                                                                                                                                                                                                                                           
  test("collateExpression: singleQuote escapes injected collation spec") {                                                                                                                                                                                                                                                                                                                                                                 
    // Before the fix singleQuote("en_US'; DROP TABLE x; --") passed through                                                                                                                                                                                                                                                                                                                                                               
    assert(singleQuote("en_US'; DROP TABLE x; --") == "'en_US''; DROP TABLE x; --'")                                                                                                                                                                                                                                                                                                                                                       
  }                                                                                                                                                                                                                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                                                                                                                                                                           
  2. subfieldExpression — the hasOnlyEscapedQuotes pass-through branch has a subtle edge case                                                                                                                                                                                                                                                                                                                                              
                                                                  
  The contract says: if the body already has only escaped (doubled) quotes, pass through verbatim. But the check is applied to the raw field string the user passed in — not to a pre-wrapped literal. This means the input "O''Brien" is treated as already-escaped and passed through, producing data['O''Brien'] which is correct. However, the input "''" (two single quotes) is also treated as already-escaped — representing the    
  empty string field name ''. That's technically valid, but extremely surprising UX. Add a comment clarifying this is intentional:
                                                                                                                                                                                                                                                                                                                                                                                                                                           
  // A body of "''" is well-formed (it represents the one-character field name "'")
  // and passes through verbatim per the documented API contract.                                                                                                                                                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                                                                                                                                                                                           
  3. isProperlyQuoted — edge case with a single ' as the entire input                                                                                                                                                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                                                                                                                                                                           
  isProperlyQuoted("'") returns false (length 1 < 2), which is correct. But isProperlyQuoted("''") returns true — it's a well-formed empty literal. Verify this is intentional: singleQuote("''") returns "''" unchanged. This means if a caller passes the empty SQL literal '', it is treated as already-quoted and emitted as-is. This is the right behavior but add a test asserting it explicitly since it's a boundary:              
                                                                  
  test("singleQuote: empty literal '' passes through unchanged") {                                                                                                                                                                                                                                                                                                                                                                         
    assert(singleQuote("''") == "''")                                                                                                                                                                                                                                                                                                                                                                                                      
  }                                                                                                                                                                                                                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                                                                                                                                                                           
  4. escapeSingleQuotes uses String.replace — correct but not obvious it replaces all occurrences                                                                                                                                                                                                                                                                                                                                          
                                                                  
  value.replace(_SingleQuote, _SingleQuote + _SingleQuote) — String.replace replaces all occurrences (not just the first, unlike regex-based replaceFirst). This is correct behavior but worth a brief note since a reader might confuse it with replaceFirst. Not a bug, just worth a comment or the existing test coverage already covers it well (it does — "a'b'c" case).                                                              
                                                                  
  5. Missing test: subfieldExpression with subfieldExpression of integer overload is untouched                                                                                                                                                                                                                                                                                                                                             
                                                                  
  subfieldExpression(expr, field: Int) is not modified and uses field directly between brackets (no quotes). This is safe since integers cannot inject SQL. But the test suite doesn't mention this variant at all — a brief note in the suite clarifying "integer overload needs no escaping" would prevent a future reader from wondering if it was missed.                                                                              
                                                                  
  6. No test for singleQuote callers in context — pattern, formatName, collationSpec                                                                                                                                                                                                                                                                                                                                                       
                                                                  
  The three real call sites of singleQuote in package.scala are:                                                                                                                                                                                                                                                                                                                                                                           
  - collateExpression (collation spec)                            
  - selectFromPathWithFormatStatement (format name, pattern)                                                                                                                                                                                                                                                                                                                                                                               
  - copyIntoTable (pattern)                                       
                             
                             
 This is generated by claude, some of them looks valid to me, please also get approval from Hemit

**Comment by sfc-gh-helmeleegy** [comment]
> 1. collateExpression call site is unprotected — same injection class
> 
> collateExpression in SqlGenerator.scala:258 calls singleQuote(collationSpec) where collationSpec comes from the user-supplied Column.collate(String) API. After this PR, singleQuote now properly escapes on the wrapping branch — so this is actually fixed as a side effect. But it's worth explicitly noting this in the PR description, and adding a test:
> 
> test("collateExpression: singleQuote escapes injected collation spec") { // Before the fix singleQuote("en_US'; DROP TABLE x; --") passed through assert(singleQuote("en_US'; DROP TABLE x; --") == "'en_US''; DROP TABLE x; --'") }
> 
> 2. subfieldExpression — the hasOnlyEscapedQuotes pass-through branch has a subtle edge case
> 
> The contract says: if the body already has only escaped (doubled) quotes, pass through verbatim. But the check is applied to the raw field string the user passed in — not to a pre-wrapped literal. This means the input "O''Brien" is treated as already-escaped and passed through, producing data['O''Brien'] which is correct. However, the input "''" (two single quotes) is also treated as already-escaped — representing the empty string field name ''. That's technically valid, but extremely surprising UX. Add a comment clarifying this is intentional:
> 
> // A body of "''" is well-formed (it represents the one-character field name "'") // and passes through verbatim per the documented API contract.
> 
> 3. isProperlyQuoted — edge case with a single ' as the entire input
> 
> isProperlyQuoted("'") returns false (length 1 < 2), which is correct. But isProperlyQuoted("''") returns true — it's a well-formed empty literal. Verify this is intentional: singleQuote("''") returns "''" unchanged. This means if a caller passes the empty SQL literal '', it is treated as already-quoted and emitted as-is. This is the right behavior but add a test asserting it explicitly since it's a boundary:
> 
> test("singleQuote: empty literal '' passes through unchanged") { assert(singleQuote("''") == "''") }
> 
> 4. escapeSingleQuotes uses String.replace — correct but not obvious it replaces all occurrences
> 
> value.replace(_SingleQuote, _SingleQuote + _SingleQuote) — String.replace replaces all occurrences (not just the first, unlike regex-based replaceFirst). This is correct behavior but worth a brief note since a reader might confuse it with replaceFirst. Not a bug, just worth a comment or the existing test coverage already covers it well (it does — "a'b'c" case).
> 
> 5. Missing test: subfieldExpression with subfieldExpression of integer overload is untouched
> 
> subfieldExpression(expr, field: Int) is not modified and uses field directly between brackets (no quotes). This is safe since integers cannot inject SQL. But the test suite doesn't mention this variant at all — a brief note in the suite clarifying "integer overload needs no escaping" would prevent a future reader from wondering if it was missed.
> 
> 6. No test for singleQuote callers in context — pattern, formatName, collationSpec
> 
> The three real call sites of singleQuote in package.scala are:
> 
> * collateExpression (collation spec)
> * selectFromPathWithFormatStatement (format name, pattern)
> * copyIntoTable (pattern)
> 
> This is generated by claude, some of them looks valid to me, please also get approval from Hemit

Thanks @sfc-gh-yuwang — addressed all six in the latest push, no behavioral changes:

1. Added collateExpression regression test exercising an injected collation spec; updated the PR description to call out the side-fix.
2. Documented the '' body edge case directly in subfieldExpression's docstring; the existing SqlInjectionSuite already asserts both branches.
3. Added the singleQuote("''") == "''" boundary test.
4. Annotated escapeSingleQuotes with a note clarifying String.replace vs replaceFirst.
5. Added an inline comment on the subfieldExpression(_, field: Int) overload + a test asserting the bare emission.
6. Added call-site tests for selectFromPathWithFormatStatement (formatName, pattern) and copyIntoTable (pattern); used .contains because the surrounding builder is whitespace-heavy.

@sfc-gh-heshah PTAL when you get a chance.

**Review by sfc-gh-heshah — CHANGES_REQUESTED** [review]
Single quoting incoming subfield expressions also needs to account for escaping backslashes and other special characters afaik.

Instead of performing string manipulation and escaping manually ourselves, have we considered using the other supported delimiter for string literals in Snowflake, `$$`? https://docs.snowflake.com/en/sql-reference/data-types-text#dollar-quoted-string-constants

If field names can be arbitrary string literals (as opposed to actual identifiers which have other constraints), I think we would be in a better position if we either start using `$$` (small blast radius), or start using bind variables for such literals (large blast radius, deeper changes in the stack).

**Comment by sfc-gh-helmeleegy** [comment]
> Single quoting incoming subfield expressions also needs to account for escaping backslashes and other special characters afaik.
> 
> Instead of performing string manipulation and escaping manually ourselves, have we considered using the other supported delimiter for string literals in Snowflake, `$$`? https://docs.snowflake.com/en/sql-reference/data-types-text#dollar-quoted-string-constants
> 
> If field names can be arbitrary string literals (as opposed to actual identifiers which have other constraints), I think we would be in a better position if we either start using `$$` (small blast radius), or start using bind variables for such literals (large blast radius, deeper changes in the stack).

Thanks @sfc-gh-heshah — adopted both: subfieldExpression and singleQuote now wrap via `$$…$$` (new dollarQuoteOrEscape helper), which closes the \' bypass at the same time. Two deliberate carve-outs though:

- Falls back to escaped single-quoting when the value contains `$$` or ends with $ (where `$$` itself would close the literal early).
- Plain strings (no ', no \) keep the pre-fix 'value' form — DataFrameWriter does string-equality on Set("'index'", "'name'") for COLUMNORDER, so unconditional `$$` broke it.

Also, made the comments a bit more concise.

File: src/main/scala/com/snowflake/snowpark/internal/analyzer/package.scala [hunk 1]
```diff
@@ -778,14 +788,74 @@ package object analyzer {
...
+          i += 2
+        } else {
+          ok = false
+        }
+      } else {
+        i += 1
+      }
     }
+    ok
   }
```
Comment (sfc-gh-heshah) [inline src/main/scala/com/snowflake/snowpark/internal/analyzer/package.scala]: Actually, isn't escaping single quotes with a backslash also a valid way to escape single quotes with a single quote delimited string literal? https://docs.snowflake.com/en/sql-reference/data-types-text#escape-sequences-in-single-quoted-string-constants

File: src/main/scala/com/snowflake/snowpark/internal/analyzer/package.scala [hunk 2]
```diff
@@ -778,14 +788,74 @@ package object analyzer {
...
+  /**
+   * True iff `value` is a well-formed Snowflake single-quoted SQL string literal: it starts and
+   * ends with `'`, and every single quote in the interior is doubled (`''`). An empty literal
+   * (`"''"`) is considered well-formed. See [[singleQuote]] for the security rationale.
+   */
+  private[analyzer] def isProperlyQuoted(value: String): Boolean =
+    value.length >= 2 &&
+      value.startsWith(_SingleQuote) &&
+      value.endsWith(_SingleQuote) &&
+      hasOnlyEscapedQuotes(value.substring(1, value.length - 1))
```
Comment (sfc-gh-heshah) [inline src/main/scala/com/snowflake/snowpark/internal/analyzer/package.scala]: I'd like this to be renamed `isStringLiteralProperlySingleQuoted` to clarify this function is specifically for delimiting string literals with single quotes.

File: src/main/scala/com/snowflake/snowpark/internal/analyzer/package.scala [hunk 3]
```diff
@@ -778,14 +788,74 @@ package object analyzer {
...
+
+  private def hasOnlyEscapedQuotes(body: String): Boolean = {
```
Comment (sfc-gh-heshah) [inline src/main/scala/com/snowflake/snowpark/internal/analyzer/package.scala]: nit: `hasOnlyEscapedSingleQuotes`

File: src/main/scala/com/snowflake/snowpark/internal/analyzer/package.scala [hunk 4]
```diff
@@ -206,8 +206,18 @@ package object analyzer {
   private[analyzer] def inExpression(column: String, values: Seq[String]): String =
     column + _In + blockExpression(values)
 
-  private[analyzer] def subfieldExpression(expr: String, field: String): String =
-    expr + _LeftBracket + _SingleQuote + field + _SingleQuote + _RightBracket
```
Comment (sfc-gh-heshah) [inline src/main/scala/com/snowflake/snowpark/internal/analyzer/package.scala:210]: There is a clear departure from the previous behavior here now that we are performing escaping in addition to quoting for the user.

Previously what it seems is we would consider the incoming `field` to be the raw unquoted body of the field name. Essentially taking backslash escaped characters and passing them through.

Now what we are doing is checking if the field has properly escaped single quotes, and if not doing some escaping of single quotes ourselves, or actually changing the behavior of the incoming field by delimiting it with double dollar signs. I'm not keen on this change after seeing the tests for it.

File: src/main/scala/com/snowflake/snowpark/internal/analyzer/package.scala [hunk 5]
```diff
@@ -778,14 +788,74 @@ package object analyzer {
...
+   * internal escape interpretation (the structural defence against SNOW-3511808). Falls back to a
+   * fully-escaped single-quoted literal via [[escapeForSingleQuotedLiteral]] when dollar-quoting is
+   * unsafe: either the value contains `$$` (which would close the literal early) or it ends with
+   * `$` (which would combine with our trailing `$$` into a malformed `$$$`).
+   */
+  private[analyzer] def dollarQuoteOrEscape(value: String): String =
+    if (value.contains("$$") || value.endsWith("$"))
+      _SingleQuote + escapeForSingleQuotedLiteral(value) + _SingleQuote
+    else
+      "$$" + value + "$$"
```
Comment (sfc-gh-heshah) [inline src/main/scala/com/snowflake/snowpark/internal/analyzer/package.scala]:
```suggestion

```
Let's remove this method entirely, I don't want to introduce a new mechanism in Snowpark that is a departure from previous behaviors.

File: src/main/scala/com/snowflake/snowpark/internal/analyzer/package.scala [hunk 6]
```diff
@@ -206,8 +206,18 @@ package object analyzer {
...
+  /**
+   * Emits `expr['field']` for variant/struct/map subfield access (SNOW-3511808). Field names that
+   * are already well-formed escaped bodies (per the `ColumnSuite.subfield` contract: every interior
+   * `'` doubled) are wrapped verbatim in single quotes; everything else flows through
+   * [[dollarQuoteOrEscape]].
+   */
+  private[analyzer] def subfieldExpression(expr: String, field: String): String = {
+    val literal =
+      if (hasOnlyEscapedQuotes(field)) _SingleQuote + field + _SingleQuote
+      else dollarQuoteOrEscape(field)
```
Comment (sfc-gh-heshah) [inline src/main/scala/com/snowflake/snowpark/internal/analyzer/package.scala]:
```suggestion
      else escapeSingleQuotesForSingleQuotedLiteral(field)
```

File: src/main/scala/com/snowflake/snowpark/internal/analyzer/package.scala [hunk 7]
```diff
@@ -778,14 +788,74 @@ package object analyzer {
...
+   *      through unchanged. 2. If `value` contains neither `'` nor `\`, emit bare `'value'` —
+   *      textually identical to the pre-fix output, which downstream string-equality checks rely on
+   *      (e.g. `DataFrameWriter`'s `columnOrder` validator on `Set("'index'", "'name'")`). 3.
+   *      Otherwise (the attack surface, including the PR #281 `\'` bypass), wrap via
+   *      [[dollarQuoteOrEscape]].
+   */
+  def singleQuote(value: String): String =
+    if (isProperlyQuoted(value)) value
+    else if (value.indexOf('\'') >= 0 || value.indexOf('\\') >= 0) dollarQuoteOrEscape(value)
+    else _SingleQuote + value + _SingleQuote
```
Comment (sfc-gh-heshah) [inline src/main/scala/com/snowflake/snowpark/internal/analyzer/package.scala]:
```suggestion
    else _SingleQuote + escapeSingleQuotesForSingleQuotedLiteral(value) + _SingleQuote
```

File: src/main/scala/com/snowflake/snowpark/internal/analyzer/package.scala [hunk 8]
```diff
@@ -778,14 +788,74 @@ package object analyzer {
...
+   * Doubles backslashes then single quotes so the value is safe to embed in a Snowflake
+   * single-quoted SQL string literal. Both are necessary because Snowflake's lexer treats `\` as an
+   * escape character (e.g. `\'` is an alternative encoding of `'`), so a quotes-only strategy is
+   * bypassable (SNOW-3511808 PR #281).
+   */
+  private[analyzer] def escapeForSingleQuotedLiteral(value: String): String =
+    value
+      .replace("\\", "\\\\")
+      .replace(_SingleQuote, _SingleQuote + _SingleQuote)
+
```
Comment (sfc-gh-heshah) [inline src/main/scala/com/snowflake/snowpark/internal/analyzer/package.scala]:
```suggestion
  private[analyzer] def escapeSingleQuotesForSingleQuotedLiteral(value: String): String =
    value
      .replace(_SingleQuote, _SingleQuote + _SingleQuote)
```

I'm rethinking the original ask to escape backslashes, because there's a lot more than just backslashes that can be escaped in string literals. Our main concern in this PR is to address potential SQL injection issues due to delimiter usage within provided strings IIUC, so I would say a more apt name and mechanism for this function would be `escapeSingleQuotesForSingleQuotedLiteral`, and only escape single quotes in the incoming string.

**Review by sfc-gh-heshah — CHANGES_REQUESTED** [review]
Sorry for the churn, I'm realizing escaping backslashes and introducing `$$` delmiting on behalf of the user is two new behaviors that entirely change how the old methods `subfieldExpression` and `singleQuote` operate.

Let's keep the changes minimal just to prevent SQL injection. Where we are dealing with a string body meant for a single quoted string literal, ensure all single quotes (the delimiters alone) are properly escaped or we do the escaping ourselves of only single quotes to prevent SQL injection.

**Comment by sfc-gh-helmeleegy** [comment]
> Sorry for the churn, I'm realizing escaping backslashes and introducing `$$` delmiting on behalf of the user is two new behaviors that entirely change how the old methods `subfieldExpression` and `singleQuote` operate.
> 
> Let's keep the changes minimal just to prevent SQL injection. Where we are dealing with a string body meant for a single quoted string literal, ensure all single quotes (the delimiters alone) are properly escaped or we do the escaping ourselves of only single quotes to prevent SQL injection.

No worries. That was a good call! I have narrowed down the PR to single-quote-only escaping. 

That said, I did like the `$$` approach because there we don't need to make any assumptions about how the Snowflake parser works, which makes it more robust to possible future changes in `'…'` parsing.

However, the cost of using `$$` is that it introduces a behavioral change for inputs containing literal backslashes (e.g., `C:\path\to\file`, `\u0041`, etc.) when those come from external sources rather than Scala source literals where the compiler has already decoded them. Snowflake interprets `\t`/`\u…` inside `'…'` literals, and `$$` would suppress that. Probably worth revisiting in a future BCR.

**Comment by sfc-gh-helmeleegy** [comment]
@sfc-gh-heshah Also, we'll need to have a new release to make this fix and the fix in https://github.com/snowflakedb/snowpark-java-scala/pull/280 available to users.

File: src/main/scala/com/snowflake/snowpark/internal/analyzer/package.scala [hunk 9]
```diff
@@ -778,14 +788,54 @@ package object analyzer {
...
+    while (ok && i < body.length) {
+      if (body.charAt(i) == '\'') {
+        if (i + 1 < body.length && body.charAt(i + 1) == '\'') {
+          i += 2
+        } else {
+          ok = false
+        }
+      } else {
+        i += 1
+      }
```
Comment (sfc-gh-heshah) [inline src/main/scala/com/snowflake/snowpark/internal/analyzer/package.scala:830]:
```suggestion
      } else if (body.charAt(i) == '\\' && i + 1 < body.length && 
                 body.charAt(i + 1) == '\'') {
        i += 2
      } else {
        i += 1
      }
```

**Review by sfc-gh-heshah — APPROVED** [review]
One minor suggestion to allow backslash escaped single quotes in `hasOnlyEscapedSingleQuotes`
